File: public/codes/sorting/countingSort.js
async function countingSort(array, setArray, setComparingIndices, setSwappingIndices, setSortedIndices, speed, isSorting) {
    const n = array.length;

    if (n === 0) {
        setSortedIndices(Array.from({ length: n }, (_, i) => i));
        setComparingIndices([]);
        return;
    }

    let max = array[0];
    for (let i = 1; i < n; i++) {
        if (!isSorting.current) return;
        setComparingIndices([i]);
        setArray([...array]);
        await sleep(speed);
        if (array[i] > max) {
            max = array[i];
        }
    }

    const count = new Array(max + 1).fill(0);
    const output = new Array(n).fill(0);

    for (let i = 0; i < n; i++) {
        if (!isSorting.current) return;
        setComparingIndices([i]);
        setArray([...array]);
        await sleep(speed);
        count[array[i]]++;
    }

    for (let i = 1; i <= max; i++) {
        if (!isSorting.current) return;
        count[i] += count[i - 1];
    }

    for (let i = n - 1; i >= 0; i--) {
        if (!isSorting.current) return;
        setComparingIndices([i]);
        setArray([...array]);
        await sleep(speed);
        output[count[array[i]] - 1] = array[i];
        count[array[i]]--;
    }

    for (let i = 0; i < n; i++) {
        if (!isSorting.current) return;
        array[i] = output[i];
        setArray([...array]);
        setSortedIndices(prev => [...prev, i]);
        await sleep(speed);
    }

    setSortedIndices(Array.from({ length: n }, (_, i) => i));
    setComparingIndices([]);
}

// Helper function (assuming it's defined elsewhere or passed)
function sleep(ms) {
    return new Promise(resolve => setTimeout(resolve, ms));
}

File: public/codes/sorting/countingSort.py
def counting_sort(array, set_array, set_comparing_indices, set_swapping_indices, set_sorted_indices, speed, is_sorting):
    n = len(array)

    if n == 0:
        set_sorted_indices(list(range(n)))
        set_comparing_indices([])
        return

    max_val = array[0]
    for i in range(1, n):
        if not is_sorting.current: return
        set_comparing_indices([i])
        set_array(list(array))
        time.sleep(speed / 1000.0)
        if array[i] > max_val:
            max_val = array[i]

    count = [0] * (max_val + 1)
    output = [0] * n

    for i in range(n):
        if not is_sorting.current: return
        set_comparing_indices([i])
        set_array(list(array))
        time.sleep(speed / 1000.0)
        count[array[i]] += 1

    for i in range(1, max_val + 1):
        if not is_sorting.current: return
        count[i] += count[i - 1]

    for i in range(n - 1, -1, -1):
        if not is_sorting.current: return
        set_comparing_indices([i])
        set_array(list(array))
        time.sleep(speed / 1000.0)
        output[count[array[i]] - 1] = array[i]
        count[array[i]] -= 1

    for i in range(n):
        if not is_sorting.current: return
        array[i] = output[i]
        set_array(list(array))
        set_sorted_indices(lambda prev: prev + [i])
        time.sleep(speed / 1000.0)

    set_sorted_indices(list(range(n)))
    set_comparing_indices([])

import time